File: project_euler/Problem079/Problem079.go
package project_euler

import (
    "bufio"
    "fmt"
    "io"
    "os"
    "strconv"
    "strings"
)

func Problem079() {
    fileName := "keylog.txt"
    length := 10
    elements := make([][]int, length)
    for i := 0; i < length; i++ {
        elements[i] = make([]int, length)
        for j := 0; j < length; j++ {
            elements[i][j] = 0
        }
    }
    visited := make([]bool, length)
    for i := 0; i < length; i++ {
        visited[i] = true
    }

    file, _ := os.Open(fileName)
    handle := bufio.NewReader(file)
    for true {
        line, _, err := handle.ReadLine()
        if io.EOF == err {
            break
        }

        value, _ := strconv.Atoi(strings.Replace(string(line), "\n", "", -1))
        createElements(value, elements, visited)
    }
    file.Close()

    digits := make([]int, 0)
    for i := 0; i < length; i++ {
        if !visited[i] {
            search(elements, i, &digits, visited, length)
        }
    }

    result := 0
    for i := len(digits) - 1; i >= 0; i-- {
        result = result * 10 + digits[i]
    }
    fmt.Println(result)
}

func createElements(value int, elements [][]int, visited []bool) {
    tail := value % 10
    visited[tail] = false
    for value >= 10 {
        value /= 10
        head := value % 10
        elements[head][tail] = 1
        tail = head
        visited[tail] = false
    }
}

func search(elements [][]int, start int, digits *[]int, visited []bool, length int) {
    visited[start] = true
    element := elements[start]
    for i := 0; i < length; i++ {
        if 0 != element[i] && !visited[i] {
            search(elements, i, digits, visited, length)
        }
    }
    *digits = append(*digits, start)
}

File: project_euler/Problem079/Problem079.py
from typing import List

class Problem079:
    def solution(self) -> None:
        file_name, length = 'keylog.txt', 10
        elements, visited = [[0 for _ in range(length)] for _ in range(length)], [True for _ in range(length)]
        with open(file_name, 'r') as handle:
            for line in handle:
                self.__create_elements(int(line.strip()), elements, visited)

        digits = []
        for i in range(length):
            if not visited[i]:
                self.__search(elements, i, digits, visited, length)

        result = 0
        for value in digits[::-1]:
            result = result * 10 + value
        print(result)

    def __create_elements(self, value: int, elements: List[List[int]], visited: List[bool]) -> None:
        tail = value % 10
        visited[tail] = False
        while value >= 10:
            value //= 10
            head = value % 10
            elements[head][tail] = 1
            tail = head
            visited[tail] = False

    def __search(self, elements: List[List[int]], start: int, digits: List[int], visited: List[bool], length: int) -> None:
        visited[start], element = True, elements[start]
        for i in range(length):
            if element[i] and not visited[i]:
                self.__search(elements, i, digits, visited, length)
        digits.append(start)
